rubicon/objc/api.py

```diff
@@ -403,29 +413,61 @@ def _objc_setter(objc_self, _cmd, new_value):
             if not isinstance(new_value, self.vartype):
                 # If vartype is a primitive, then new_value may be unboxed. If that is the case, box it manually.
                 new_value = self.vartype(new_value)
-            old_value = get_ivar(objc_self, '_' + attr_name)
-            if issubclass(self.vartype, objc_id) and new_value:
+
+            if issubclass(self.vartype, objc_id) and not self.weak:
+                old_value = get_ivar(objc_self, ivar_name, weak=self.weak)
+
+                if new_value.value == old_value.value:
+                    # old and new value are the same, nothing to do
+                    return
+
+            if not self.weak and issubclass(self.vartype, objc_id) and new_value:
                 # If the new value is a non-null object, retain it.
                 send_message(new_value, 'retain', restype=objc_id, argtypes=[])
-            set_ivar(objc_self, '_' + attr_name, new_value)
-            if issubclass(self.vartype, objc_id) and old_value:
+
+            set_ivar(objc_self, ivar_name, new_value, weak=self.weak)
+
+            if not self.weak and issubclass(self.vartype, objc_id) and old_value:
                 # If the old value is a non-null object, release it.
                 send_message(old_value, 'release', restype=None, argtypes=[])
 
         setter_name = 'set' + attr_name[0].upper() + attr_name[1:] + ':'
 
-        add_method(
+        replace_method(
             class_ptr, attr_name, _objc_getter,
             [self.vartype, ObjCInstance, SEL],
         )
-        add_method(
+        replace_method(
             class_ptr, setter_name, _objc_setter,
             [None, ObjCInstance, SEL, self.vartype],
         )
 
         attrs = self._get_property_attributes()
         libobjc.class_addProperty(class_ptr, ensure_bytes(attr_name), attrs, len(attrs))
 
+        # Add cleanup routines to dealloc.
+
+        old_dealloc = libobjc.class_getMethodImplementation(class_ptr, SEL("dealloc"))
+
+        def _new_delloc(objc_self, _cmd):
+
+            # Clean up ivar.
+            if self.weak:
+                # Clean up weak reference.
+                ivar = libobjc.class_getInstanceVariable(libobjc.object_getClass(objc_self), ivar_name.encode())
+                libobjc.objc_storeWeak(objc_self.value + libobjc.ivar_getOffset(ivar), None)
+            elif issubclass(self.vartype, objc_id):
+                # If the old value is a non-null object, release it. There is no need to set the actual ivar to nil.
+                old_value = get_ivar(objc_self, ivar_name, weak=self.weak)
+                send_message(old_value, 'release', restype=None, argtypes=[])
```

Comment:
Could this be simplified somehow by calling `_objc_setter` or `set_ivar`? Especially in the weak case the code is a bit more complex, so it would be better to use our existing functions instead of reimplementing it.

Reply:
Sure.

Reply:
I remember now why I did this. The current `set_ivar` implementation raises a TypeError when setting the value to `None`. This can be changed of course, it is common place in ObjC to set a value to nil.